// http-auth/gensrc/auth/basic.js
// Generated by CoffeeScript 1.10.0
(function() {
  var Base, Basic, htpasswd, utils,
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  Base = require('./base');

  utils = require('./utils');

  htpasswd = require('htpasswd');

  Basic = (function(superClass) {
    extend(Basic, superClass);

    function Basic(options1, checker1) {
      this.options = options1;
      this.checker = checker1;
      Basic.__super__.constructor.call(this, this.options, this.checker);
    }

    Basic.prototype.processLine = function(line) {
      var hash, lineSplit, username;
      lineSplit = line.split(":");
      username = lineSplit.shift();
      hash = lineSplit.join(":");
      return this.options.users.push({
        username: username,
        hash: hash
      });
    };

    Basic.prototype.findUser = function(req, hash, callback) {
      var found, i, len, password, ref, splitHash, user, username;
      splitHash = (utils.decodeBase64(hash)).split(":");
      username = splitHash.shift();
      password = splitHash.join(":");
      if (this.checker) {
        return this.checker.apply(this, [
          username, password, (function(_this) {
            return function(success) {
              return callback.apply(_this, [
                {
                  user: success ? username : void 0
                }
              ]);
            };
          })(this)
        ]);
      } else {
        ref = this.options.users;
        for (i = 0, len = ref.length; i < len; i++) {
          user = ref[i];
          if (user.username === username && htpasswd.verify(user.hash, password)) {
            found = true;
            break;
          }
        }
        return callback.apply(this, [
          {
            user: found ? username : void 0
          }
        ]);
      }
    };

    Basic.prototype.parseAuthorization = function(header) {
      var hash, ref, type;
      ref = header.split(" "), type = ref[0], hash = ref[1];
      if (type === "Basic") {
        return hash;
      }
    };

    Basic.prototype.generateHeader = function() {
      return "Basic realm=\"" + this.options.realm + "\"";
    };

    return Basic;

  })(Base);

  module.exports = function(options, checker) {
    return new Basic(options, checker);
  };

}).call(this);
